backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/boxing/TypeParameterEliminator.kt

```diff
@@ -0,0 +1,148 @@
+        private val typeParametersMapping: MutableMap<IrTypeParameterSymbol, IrType>,
+        typeRemapper: TypeRemapper = TypeRemapperWithParametersElimination(symbolRemapper, typeParametersMapping)
+) : DeepCopyIrTreeWithSymbols(symbolRemapper, typeRemapper, symbolRenamer) {
+
+    init {
+        // TODO refactor
+        (typeRemapper as? TypeRemapperWithParametersElimination)?.let {
+            it.deepCopy = this
+        }
+    }
+
+    fun addMapping(typeParameter: IrTypeParameter, concreteType: IrType) {
+        typeParametersMapping[typeParameter.symbol] = concreteType
+    }
+
+    private fun <E : IrTypeParametersContainer> E.withEliminatedTypeParameters(): E = apply {
+        typeParameters.removeIf { it.symbol in typeParametersMapping }
+    }
+
+    /**
+     * After elimination, it may turn out that this call satisfies condition for the callee to get specialization.
+     *
+     * For example:
+     *
+     * ```
+     * fun <T> id(x: T) = x
+     * fun <U> id2(y: U) = id(y)
+     * ...
+     * id2(42)
+     * ```
+     *
+     * After elimination of type parameter U in `id2-Int` it turns out that `id(y)` is a call to generic function
+     * with primitive type argument, so it makes sense to specialize function `id` too, recursively.
+     */
+    override fun visitCall(expression: IrCall): IrCall {
+        val newCall = if (expression.satisfiesSpecializationCondition(typeParametersMapping)) {
+            symbolRemapper.inIgnoreFunctionMappingMode(expression.symbol.owner) { super.visitCall(expression) }
+        } else {
+            super.visitCall(expression)
+        }
+        return newCall.transform(specializationTransformer, null) as IrCall
+    }
+
+    override fun visitSimpleFunction(declaration: IrSimpleFunction): IrSimpleFunction {
+        return super.visitSimpleFunction(declaration).withEliminatedTypeParameters().apply {
+            // Little crutch -- assuming that function does not have parent iff it's new (e.g. a specialization).
+            // TODO elaborate on it later
+            try {
+                parent
+            } catch (_: UninitializedPropertyAccessException) {
+                parent = declaration.parent
+            }
+            acceptVoid(SafeWrappedDescriptorPatcher)
+            if (descriptor is WrappedPropertyGetterDescriptor) {
+                this.correspondingPropertySymbol = declaration.correspondingPropertySymbol
+            }
+        }
+    }
+
+    private fun IrCall.satisfiesSpecializationCondition(typeParameterMapping: MutableMap<IrTypeParameterSymbol, IrType>): Boolean {
+        if (typeArgumentsCount != 1) return false
+        val symbol = getTypeArgument(0)?.classifierOrNull
+        return symbol in typeParameterMapping && typeParameterMapping[symbol]?.isPrimitiveType() == true
+    }
+}
+
+/**
+ * Copy of [DeepCopyTypeRemapper] with additional remapping for eliminated type parameters.
+ * TODO provide delegate
+ */
+class TypeRemapperWithParametersElimination(
```

Comment:
Take a look at `DeepCopyIrTreeWithDescriptors.kt` in Kotlin repository. There is class `InlinerTypeRemapper` over there, it does similar work.